runners/maven-plugin/build.gradle.kts

```diff
@@ -1,95 +1,101 @@
-        }
-        filter { line ->
-            line.replace(
-                "<version>maven-plugin-plugin</version>",
-                "<version>${setupMaven.mavenPluginToolsVersion}</version>"
-            )
-        }
+    val pomTemplateFile = layout.projectDirectory.file("pom.tpl.xml")
+
+    from(pomTemplateFile) {
+        rename { it.replace(".tpl.xml", ".xml") }
+
+        expand(
+            "mavenVersion" to setupMavenProperties.mavenVersion,
+            "dokka_version" to dokka_version,
+            "mavenPluginToolsVersion" to setupMavenProperties.mavenPluginToolsVersion,
+        )
     }
+
+    into(temporaryDir)
 }
 
-val syncClasses by tasks.registering(Sync::class) {
-    description = "Copy compiled classes to the Maven build dir, for Maven Plugin task execution"
+val prepareMavenPluginBuildDir by tasks.registering(Sync::class) {
+    description = "Prepares all files for Maven Plugin task execution"
+    group = mavenPluginTaskGroup
 
-    dependsOn(tasks.compileKotlin, tasks.compileJava)
-    from("$buildDir/classes/kotlin", "$buildDir/classes/java")
-    into("${setupMaven.mavenBuildDir}/classes/java")
+    from(tasks.compileKotlin.flatMap { it.destinationDirectory }) { into("classes/java/main") }
+    from(tasks.compileJava.flatMap { it.destinationDirectory }) { into("classes/java/main") }
 
-    preserve {
-        include("**/*.class")
-    }
+    from(generatePom)
+
+    into(setupMavenProperties.mavenBuildDir)
 }
 
 val helpMojo by tasks.registering(CrossPlatformExec::class) {
-    dependsOn(setupMaven, generatePom, syncClasses)
-    workingDir(setupMaven.mavenBuildDir)
-    commandLine(setupMaven.mvn, "-e", "-B", "org.apache.maven.plugins:maven-plugin-plugin:helpmojo")
-
-    outputs.dir(layout.buildDirectory.dir("maven"))
+    group = mavenPluginTaskGroup
+
+    dependsOn(tasks.installMavenBinary, prepareMavenPluginBuildDir)
+    workingDir(setupMavenProperties.mavenBuildDir)
+    executable(setupMavenProperties.mvn.get())
+    args("-e", "-B", "org.apache.maven.plugins:maven-plugin-plugin:helpmojo")
+
+    outputs.dir(setupMavenProperties.mavenBuildDir)
+
+    doLast("normalize maven-plugin-help.properties") {
+        // The maven-plugin-help.properties file contains a timestamp by default.
+        // It should be removed as it is not reproducible and impacts Gradle caching
+        val pluginHelpProperties = workingDir.resolve("maven-plugin-help.properties")
+        pluginHelpProperties.writeText(
+            buildString {
+                val lines = pluginHelpProperties.readText().lines().iterator()
+                // the first line is a descriptive comment
+                appendReproducibleNewLine(lines.next())
+                // the second line is the timestamp, which should be ignored
+                lines.next()
+                // the remaining lines are properties
+                lines.forEach { appendReproducibleNewLine(it) }
+            }
+        )
+    }
```

Comment:
Unfortunately the generated `maven-plugin-help.properties` contains a timestamp, so Gradle will always consider the output out-of-date.

I wrote this code to fix it - but maybe there's a Maven flag that will disable the timestamp?

This is perhaps a bit OTT/pre-optimising.

Reply:
I actually don't understand what helpMojo is supposed to achieve, and I can't google anything sensible. It looks like it should either generate the `help` goal (which Dokka's Maven plugin doesn't have) or maybe add some help documentation? Were you able to figure it out?

Nothing looks broken to me, so I think it's fine to ignore the timestamp. Perhaps, at some point our Maven plugin should be revisited as well :sweat_smile:

Reply:
I think it auto generates a 'help' command based on the properties in the Maven plugin

The file it generates is in `runners/maven-plugin/build/maven/generated-sources`

But, like you, I'm not sure! I've never developed a Maven plugin.

```java

package org.jetbrains.dokka.maven;

/**
 * Display help information on dokka-maven-plugin.<br>
 * Call <code>mvn dokka:help -Ddetail=true -Dgoal=&lt;goal-name&gt;</code> to display parameter details.
 * @author maven-plugin-tools
 * @goal help
 * @requiresProject false
 * @threadSafe
 */
public class HelpMojo
    extends AbstractMojo
{
    /**
     * If <code>true</code>, display all settable properties for each goal.
     *
     * @parameter property="detail" default-value="false"
     */
    private boolean detail;

    /**
     * The name of the goal for which to show help. If unspecified, all goals will be displayed.
     *
     * @parameter property="goal"
     */
    private java.lang.String goal;

    /**
     * The maximum length of a display line, should be positive.
     *
     * @parameter property="lineLength" default-value="80"
     */
    private int lineLength;

...
```

I was looking at other options, and I found this Gradle plugin: https://github.com/britter/maven-plugin-development. It might help with developing a Maven plugin, but on the other hand, the current approach seems to be working.